### PR TITLE
[SPARK-12941][SQL][MASTER] Spark-SQL JDBC Oracle dialect fails to map string datatypes to Oracle VARCHAR datatype

### PR DESCRIPTION
**Pull request by thomastechs**
## What changes were proposed in this pull request?

This Pull request is used for the fix SPARK-12941, creating a data type mapping to Oracle for the corresponding data type"Stringtype" from dataframe. This PR is for the master branch fix, where as another PR is already tested with the branch 1.4
## How was the this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
This patch was tested using the Oracle docker .Created a new integration suite for the same.The oracle.jdbc jar was to be downloaded from the maven repository.Since there was no jdbc jar available in the maven repository, the jar was downloaded from oracle site manually and installed in the local; thus tested. So, for SparkQA test case run, the ojdbc jar might be manually placed in the local maven repository(com/oracle/ojdbc6/11.2.0.2.0) while Spark QA test run.

### CODE REVIEW

**Comment by yhuai** [comment]
ok to test

**Comment by yhuai** [comment]
@thomastechs Thank you for the PR. Since the jar is not in maven repo and you have tested it locally, how about we ignore the test? Let's keep the test but just ignore it. Let's also comment how to run the test. What do you think?

**Comment by yhuai** [comment]
ok to test

**Comment by thomastechs** [comment]
@yhuai  Thanks for the comments. I have put the ignore instead of test. The steps for the testing would be as follows.
1. Pull oracle 11g image - docker pull wnameless/oracle-xe-11g
2. Start docker - sudo service docker start
3. Download oracle 11g driver jar and put it in maven local repo. (com/oracle/ojdbc6/11.2.0.2.0/ojdbc6-11.2.0.2.0.jar)
4. Run spark test - ./build/sbt "test-only org.apache.spark.sql.jdbc.OracleIntegrationSuite"

**Comment by thomastechs** [comment]
@yhuai  The style checks have been passed. Currently the build fails because the maven dependency oracle jar is not available, as expected. Suppose, we comment the oracle jar dependency in the pom file, still the build might fail in the beforeAll step where it tries to create a connection to Oracle docker using the ojdbc jar(which is not present in SparkQA env). So what would you suggest to proceed?

File: docker-integration-tests/pom.xml [hunk 1]
```diff
@@ -131,6 +131,12 @@
       <artifactId>postgresql</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.oracle</groupId>
+      <artifactId>ojdbc6</artifactId>
+      <version>11.2.0.2.0</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (yhuai) [inline docker-integration-tests/pom.xml]: Can you remove this change?

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala [hunk 2]
```diff
@@ -112,7 +112,7 @@ abstract class DockerJDBCIntegrationSuite
       // Start the container and wait until the database can accept JDBC connections:
       docker.startContainer(containerId)
       jdbcUrl = db.getJdbcUrl(dockerIp, externalPort)
-      eventually(timeout(60.seconds), interval(1.seconds)) {
+      eventually(timeout(200.seconds), interval(200.seconds)) {
```
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala]: do we need to change this line?

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import java.math.BigDecimal
+import java.sql.{Connection, Date, Timestamp}
+import java.util.Properties
+
+import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.sql.types._
+import org.apache.spark.tags.DockerTest
+
+@DockerTest
+class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLContext {
```
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Please add doc to explain how to run this suite.

Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Also, please add doc to explain why we have to ignore all tests of this suite.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala (hunk 2 above)
Comment (thomastechs) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala]: @yhuai  This line of timeout is required for the oracle related test.Because, the time taken by the oracle instance to bring up would not be enough with the old setting. We tested it in our local and  
confirmed this. If we do not change this time duration, Oracle would throw an exception, saying that the instance was tried to hit during the initialization stage.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala (hunk 3 above)
Comment (thomastechs) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: @yhuai  So, I shall add the comments in the OracleIntegrationSuite.scala itself on how to run ,version of ojdbc jar, and on why ignore. Hope this is what you meant by doc.

**Comment by thomastechs** [comment]
retest please. ok to test

**Comment by yhuai** [comment]
test this please

**Comment by yhuai** [comment]
test this please

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala (hunk 2 above)
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala]: Seems we need to have a smaller number of `interval`. Right now, the number of internal and timeout are the same.

File: docker-integration-tests/pom.xml [hunk 4]
```diff
@@ -131,6 +131,14 @@
...
       <scope>test</scope>
     </dependency>
+    <!--
+      <dependency>
+        <groupId>com.oracle</groupId>
+        <artifactId>ojdbc6</artifactId>
+        <version>11.2.0.2.0</version>
+        <scope>test</scope>
+      </dependency>
+    -->
```
Comment (yhuai) [inline docker-integration-tests/pom.xml]: Please also add comment at here to explain why we comment out this dependency.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+package org.apache.spark.sql.jdbc
+
+import java.math.BigDecimal
+import java.sql.{Connection, Date, Timestamp}
+import java.util.Properties
+
+import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.sql.types._
+import org.apache.spark.tags.DockerTest
```
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Add a newline after this line.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * This integration suite is created as part of Pull request(11306) and is used for the fix
+ * SPARK-12941, creating a data type mapping to Oracle for the corresponding data type
+ * "Stringtype" from dataframe. This PR is for the master branch fix, where as another PR
+ * is already tested with the branch 1.4
+ *
+ * This patch was tested using the Oracle docker.Created this integration suite for the same.
```
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Whitespace before `Created`.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Properties
+
+import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.sql.types._
+import org.apache.spark.tags.DockerTest
+/**
+ * This integration suite is created as part of Pull request(11306) and is used for the fix
+ * SPARK-12941, creating a data type mapping to Oracle for the corresponding data type
+ * "Stringtype" from dataframe. This PR is for the master branch fix, where as another PR
+ * is already tested with the branch 1.4
```
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Seems we do not need to have this paragraph.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ * while Spark QA test run
+ *
+ * The following would be the steps to test this
+ * 1. Pull oracle 11g image - docker pull wnameless/oracle-xe-11g
+ * 2. Start docker - sudo service docker start
+ * 3. Download oracle 11g driver jar and put it in maven local repo:
+ *    (com/oracle/ojdbc6/11.2.0.2.0/ojdbc6-11.2.0.2.0.jar)
+ * 4. Run spark test - ./build/sbt "test-only org.apache.spark.sql.jdbc.OracleIntegrationSuite"
+ *
+ */
```
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: We can mention that all tests in this suite are ignored and then delete the comment above the test case below.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala (hunk 2 above)
Comment (thomastechs) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala]: @yhuai    interval to be set as 200 sec, was a requirement to execute it fine in our local environment of oracle on docker.Less than that it causes exception related to early hit. Since the current test case is ignored, we may revert it back to 60,1; adding a comment here saying to increase this number, when testing with Oracle. Please let me know your thoughts.

Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala]: Yea, sounds good. Let's revert this change and add a comment to `OracleIntegrationSuite`.

Comment (thomastechs) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala]: Sure; A build is already in progress, I guess I can commit after that

**Comment by jayadevanmurali** [comment]
@SparkQA Seems to be the build failed due kmeans error. I think this is not related to this PR. Could you please check

**Comment by thomastechs** [comment]
@yhaui @SparkQA Looks like the build failed due to some other unit test cases, not related to this PR
Could you please confirm?

**Comment by thomastechs** [comment]
@yhuai  Could you initiate a retest please?

**Comment by yhuai** [comment]
test this please

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  import testImplicits._
+
+  override val db = new DatabaseOnDocker {
+    override val imageName = "wnameless/oracle-xe-11g:latest"
+    override val env = Map(
+      "ORACLE_ROOT_PASSWORD" -> "oracle"
+    )
+    override val jdbcPort: Int = 1521
+    override def getJdbcUrl(ip: String, port: Int): String =
+      s"jdbc:oracle:thin:system/oracle@//$ip:$port/xe"
```
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: Sorry. Last question. If we leave these at here, we are still trying to establish the connection to the database through JDBC, right? Since we do not have jdbc driver, it will fail. Should we just comment out these code or just comment out the class?

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+ *    (com/oracle/ojdbc6/11.2.0.2.0/ojdbc6-11.2.0.2.0.jar)
+ * 4. The timeout and interval parameter to be increased from 60,1 to a high value for oracle test
+ *    in DockerJDBCIntegrationSuite.scala (Locally tested with 200,200 and executed successfully).
+ * 5. Run spark test - ./build/sbt "test-only org.apache.spark.sql.jdbc.OracleIntegrationSuite"
+ *
+ * All tests in this suite are ignored because of the dependency with the oracle jar from maven
+ * repository.
+ */
+@DockerTest
+class OracleIntegrationSuite extends DockerJDBCIntegrationSuite with SharedSQLContext {
```
Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: How about we use `Ignore` tag to ignore the entire suite? So, we should not even try to connect to oracle.

Comment (yhuai) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: oh nvm. Before will not be triggered since no test will run.

**Comment by yhuai** [comment]
Thanks. Merging to master.

File: docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala (hunk 10 above)
Comment (jayadevanmurali) [inline docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/OracleIntegrationSuite.scala]: @yhuai the beforeall method will not execute because there is no valid test inside this class(already put ignore on top of test).
